peer/network.go

```diff
@@ -366,10 +357,7 @@ func (n *network) AppRequest(ctx context.Context, nodeID ids.NodeID, requestID u
 // Error returned by this function is expected to be treated as fatal by the engine
 // If [requestID] is not known, this function will emit a log and return a nil error.
 // If the response handler returns an error it is propagated as a fatal error.
-func (n *network) AppResponse(_ context.Context, nodeID ids.NodeID, requestID uint32, response []byte) error {
-	n.lock.Lock()
-	defer n.lock.Unlock()
-
+func (n *network) AppResponse(ctx context.Context, nodeID ids.NodeID, requestID uint32, response []byte) error {
 	if n.closed.Get() {
```

Comment:
This is a very subtle change. So we should be terrified to make it.

I believe we _must_ hold the lock when checking for `n.closed` on all of the inbound `response` + `requestFailed` flows to avoid a potential panic due to a write to a closed channel (or closing a channel twice).

Reply:
As discussed offline I'll see if I can make a separate regression test for this invariant as a follow-up to this PR.

Reply:
I documented the invariant on `n.closed` as well 👍